### PR TITLE
Add util methods

### PR DESCRIPTION
**Pull request by bochengchu**
Adding GetBoolean and GetStruct methods in support of nested JWT.

We need this in https://github.com/envoyproxy/envoy/pull/23444

### CODE REVIEW

**Comment by aryan16** [comment]
@cc : qiwzhang

**Comment by aryan16** [comment]
cc : @qiwzhang

File: src/struct_utils.cc [hunk 1]
```diff
@@ -51,6 +51,40 @@ StructUtils::FindResult StructUtils::GetUInt64(const std::string& name,
...
+                                                bool* value) {
+  const auto& fields = struct_pb_.fields();
+  const auto it = fields.find(name);
+  if (it == fields.end()) {
+    return MISSING;
+  }
+  if (it->second.kind_case() != google::protobuf::Value::kBoolValue) {
+    return WRONG_TYPE;
+  }
+  if (it->second.number_value() < 0) {
```
Comment (qiwzhang) [inline src/struct_utils.cc:64]: you should use bool_value()

File: src/struct_utils.cc [hunk 2]
```diff
@@ -51,6 +51,40 @@ StructUtils::FindResult StructUtils::GetUInt64(const std::string& name,
...
+    return NOT_POSITIVE;
+  }
+  *value = static_cast<bool>(it->second.bool_value());
```
Comment (qiwzhang) [inline src/struct_utils.cc:67]: do we need static_cast?   bool_value() is bool

File: src/struct_utils.cc [hunk 3]
```diff
@@ -51,6 +51,40 @@ StructUtils::FindResult StructUtils::GetUInt64(const std::string& name,
...
+  if (it == fields.end()) {
+    return MISSING;
+  }
+  if (it->second.kind_case() != google::protobuf::Value::kStructValue) {
+    return WRONG_TYPE;
+  }
+  if (it->second.number_value() < 0) {
+    return NOT_POSITIVE;
+  }
+  *value = static_cast<google::protobuf::Struct>(it->second.struct_value());
```
Comment (qiwzhang) [inline src/struct_utils.cc:84]: This is a proto.Struct copy,  are you sure you need to copy it out?

File: src/struct_utils.cc [hunk 4]
```diff
@@ -51,6 +51,40 @@ StructUtils::FindResult StructUtils::GetUInt64(const std::string& name,
...
+  const std::string& name, google::protobuf::Struct* value) {
+  const auto& fields = struct_pb_.fields();
+  const auto it = fields.find(name);
+  if (it == fields.end()) {
+    return MISSING;
+  }
+  if (it->second.kind_case() != google::protobuf::Value::kStructValue) {
+    return WRONG_TYPE;
+  }
+  if (it->second.number_value() < 0) {
```
Comment (qiwzhang) [inline src/struct_utils.cc:81]: should not use number_value

File: src/struct_utils.cc (hunk 2 above)
Comment (bochengchu) [reply src/struct_utils.cc]: Fixed

File: src/struct_utils.cc (hunk 3 above)
Comment (bochengchu) [reply src/struct_utils.cc]: Removed static_cast.

File: test/jwt_test.cc [hunk 5]
```diff
@@ -444,6 +444,53 @@ TEST(JwtParseTest, InvalidSignature) {
             Status::JwtSignatureParseErrorBadBase64);
 }
 
+TEST(JwtParseTest, GoodNestedJwt) {
+  const std::string jwt_text =
```
Comment (qiwzhang) [inline test/jwt_test.cc:448]: For easy to read,  please add comment to list encoded claim body.

File: test/jwt_test.cc [hunk 6]
```diff
@@ -444,6 +444,53 @@ TEST(JwtParseTest, InvalidSignature) {
...
+  StructUtils payload_getter_1(jwt.payload_pb_);
+  ::google::protobuf::Struct struct_value;
+  EXPECT_EQ(payload_getter_1.GetStruct("nested", &struct_value),
+            StructUtils::OK);
+  StructUtils payload_getter_2(struct_value);
+  std::string string_value;
+  EXPECT_EQ(payload_getter_2.GetString("key-1", &string_value),
+            StructUtils::OK);
+  EXPECT_EQ(string_value, "value1");
+  ::google::protobuf::Struct nested_struct_value;
```
Comment (qiwzhang) [inline test/jwt_test.cc:479]: add blank line above 479

File: test/jwt_test.cc [hunk 7]
```diff
@@ -444,6 +444,53 @@ TEST(JwtParseTest, InvalidSignature) {
...
+
+  StructUtils payload_getter_1(jwt.payload_pb_);
+  ::google::protobuf::Struct struct_value;
+  EXPECT_EQ(payload_getter_1.GetStruct("nested", &struct_value),
+            StructUtils::OK);
+  StructUtils payload_getter_2(struct_value);
+  std::string string_value;
+  EXPECT_EQ(payload_getter_2.GetString("key-1", &string_value),
+            StructUtils::OK);
+  EXPECT_EQ(string_value, "value1");
```
Comment (qiwzhang) [inline test/jwt_test.cc:478]: add a comment at 478 as:
```
// fetching: nested.key-1 = value1
```
do the same for other value fetching

**Review by qiwzhang — APPROVED** [review]

**Pull request merged**